File: frontend/src/types/question.ts
export type Difficulty = 'easy' | 'medium' | 'hard'
export type OptionKey = 'A' | 'B' | 'C' | 'D'

export interface Option {
  id: string
  option_key: OptionKey
  text: string
  image_url: string | null
}

export interface OptionInput {
  option_key: OptionKey
  text: string
  image_url?: string | null
  is_correct: boolean
}

export interface Question {
  id: string
  subject_id: string
  topic_id: string | null
  text: string
  image_url: string | null
  difficulty: Difficulty
  explanation: string | null
  tags: string[]
  source_doc_id: string | null
  created_at: string
  options: Option[]
}

export interface QuestionCreatePayload {
  subject_id: string
  text: string
  difficulty: Difficulty
  options: OptionInput[]
  topic_id?: string | null
  image_url?: string | null
  explanation?: string | null
  tags?: string[]
}

export interface QuestionUpdatePayload {
  text?: string
  image_url?: string | null
  difficulty?: Difficulty
  explanation?: string | null
  tags?: string[]
}

export interface QuestionFilters {
  subject_id?: string
  topic_id?: string
  difficulty?: Difficulty
  page?: number
  limit?: number
}

File: backend/app/schemas/question.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

Difficulty = Literal["easy", "medium", "hard"]
OptionKey = Literal["A", "B", "C", "D"]


# ─── Option ─────────────────────────────────────────────
class OptionCreate(BaseModel):
    option_key: OptionKey
    text: str
    image_url: str | None = None
    is_correct: bool


class OptionOut(BaseModel):
    """Teacher/admin-facing option. Note: omits `is_correct` to match the
    contract's OptionOut shape — it is never revealed in API responses."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    option_key: OptionKey
    text: str
    image_url: str | None = None


# ─── Question ───────────────────────────────────────────
class QuestionCreate(BaseModel):
    subject_id: UUID
    text: str
    difficulty: Difficulty
    options: list[OptionCreate] = Field(min_length=4, max_length=4)
    topic_id: UUID | None = None
    image_url: str | None = None
    explanation: str | None = None
    tags: list[str] = []


class QuestionUpdate(BaseModel):
    text: str | None = None
    image_url: str | None = None
    difficulty: Difficulty | None = None
    explanation: str | None = None
    tags: list[str] | None = None


class QuestionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    subject_id: UUID
    topic_id: UUID | None = None
    text: str
    image_url: str | None = None
    difficulty: Difficulty
    explanation: str | None = None
    tags: list[str] = []
    source_doc_id: UUID | None = None
    created_at: datetime
    options: list[OptionOut] = []
